dhis-2/dhis-api/src/main/java/org/hisp/dhis/keyjsonvalue/KeyJsonValueQuery.java

```diff
@@ -0,0 +1,191 @@
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.keyjsonvalue;
+
+import static java.lang.Character.isLetterOrDigit;
+import static java.util.Collections.emptyList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.ToString;
+
+import org.hisp.dhis.common.NamedParams;
+
+/**
+ * Details of a query as it can be performed to fetch {@link KeyJsonValue}s.
+ *
+ * @author Jan Bernitt
+ */
+@ToString
+@Getter
+@Builder( toBuilder = true )
+@AllArgsConstructor( access = AccessLevel.PRIVATE )
+public final class KeyJsonValueQuery
+{
+
+    private final String namespace;
+
+    /**
+     * By default, only entries which have at least one non-null value for the
+     * extracted fields are returned. If all are included even matches with only
+     * null values are included in the result list.
+     */
+    private final boolean includeAll;
+
+    @Builder.Default
+    private final List<Field> fields = emptyList();
+
+    @ToString
+    @Getter
+    public static final class Field
+    {
+        private final String path;
+
+        private final String alias;
+
+        public Field( String path )
+        {
+            this( path, null );
+        }
+
+        public Field( String path, String alias )
+        {
+            if ( path == null || !path.matches( "^[-_a-zA-Z0-9]{1,32}(?:\\.[-_a-zA-Z0-9]{1,32}){0,5}$" ) )
```

Comment:
Factor out, centralize and document the regex.

Reply:
Good point. I wanted to point out that this also has security reasons. Will do that in my comment on the constant I extract.

Reply:
done